File: BMWCenter/Core/Care/CareTypes.swift
import Foundation

enum CueSeverity: String, Sendable, Comparable {
    case celebration
    case coach
    case protective
    case critical

    private var rank: Int {
        switch self {
        case .celebration: return 0
        case .coach: return 1
        case .protective: return 2
        case .critical: return 3
        }
    }

    static func < (lhs: CueSeverity, rhs: CueSeverity) -> Bool {
        lhs.rank < rhs.rank
    }
}

struct CareCue: Identifiable, Sendable, Equatable {
    let id: String
    let text: String
    let severity: CueSeverity
    let localizationKey: String
    var chipTitle: String?

    init(
        id: String,
        text: String,
        severity: CueSeverity,
        localizationKey: String = "",
        chipTitle: String? = nil
    ) {
        self.id = id
        self.text = text
        self.severity = severity
        self.localizationKey = localizationKey.isEmpty ? id : localizationKey
        self.chipTitle = chipTitle
    }
}

struct CareChannelPlan: Sendable, Equatable {
    var speak: Bool
    var carPlayRow: Bool
    var carPlayAlert: Bool
    var phoneChip: Bool
    var fullScreen: Bool
    var notification: Bool
    var haptic: String?
    var toneCount: Int
}

struct CareContext: Sendable {
    var now: Date = .now
    var supportedPIDs: Set<UInt8> = []
    var ambientC: Double?
    var oilTempC: Double?
    var oilIsEstimated: Bool = true
    var engineReady: Bool = false
    var readiness: Double = 0
    var isColdPhase: Bool = true
    var isCityTraffic: Bool = false
    var tripDistanceKm: Double = 0
    var tripDurationS: Double = 0
    var isVehicleStopped: Bool = true
    var fuelType: FuelType = .gasoline
    var isTurbo: Bool = true
    var sensitivityOffsetC: Double = 0
    var liveEcoScore: Double = 100
    var thermalCountdownS: Double?
    var lastRefuelAt: Date?
}

enum CareBucket {
    static func ambient(_ c: Double?) -> String {
        guard let c else { return "ambient:unknown" }
        switch c {
        case ..<0: return "ambient:<0"
        case 0..<10: return "ambient:0-10"
        case 10..<20: return "ambient:10-20"
        case 20..<30: return "ambient:20-30"
        default: return "ambient:>30"
        }
    }

    static func load(_ pct: Double?) -> String {
        guard let pct else { return "load:unknown" }
        switch pct {
        case ..<40: return "load:<40"
        case 40..<70: return "load:40-70"
        default: return "load:>70"
        }
    }

    static func speed(_ kmh: Double?) -> String {
        guard let kmh else { return "speed:unknown" }
        switch kmh {
        case ..<0.5: return "speed:0"
        case 0.5..<50: return "speed:1-50"
        case 50..<90: return "speed:50-90"
        default: return "speed:>90"
        }
    }

    static func composite(ambient: Double?, load: Double?, speed: Double?) -> String {
        [self.ambient(ambient), self.load(load), self.speed(speed)].joined(separator: "|")
    }
}

enum LocalizedCare {
    static func string(_ key: String, _ args: CVarArg...) -> String {
        let format = String(localized: String.LocalizationValue(key), table: "Localizable")
        guard !args.isEmpty else { return format }
        return String(format: format, locale: .current, arguments: args)
    }
}

File: mobile/src/core/care/careTypes.ts
import { archetypeProfile, type VehicleDiagnosticProfile } from "../vehicle/vehicleProfile";

export type CueSeverity = "celebration" | "coach" | "protective" | "critical";

const SEVERITY_RANK: Record<CueSeverity, number> = {
  celebration: 0,
  coach: 1,
  protective: 2,
  critical: 3,
};

export function severityRank(s: CueSeverity): number {
  return SEVERITY_RANK[s];
}

export interface CareCue {
  id: string;
  text: string;
  severity: CueSeverity;
  localizationKey?: string;
}

export interface CareChannelPlan {
  speak: boolean;
  phoneChip: boolean;
  fullScreen: boolean;
  notification: boolean;
  toneCount: number;
}

export interface CareContext {
  now: number;
  /** Per-vehicle thermal/electrical constants. Watchdogs must judge against this
   * rather than one hardcoded threshold for every car. */
  vehicle: VehicleDiagnosticProfile;
  ambientC?: number;
  oilTempC?: number;
  oilIsEstimated: boolean;
  engineReady: boolean;
  isColdPhase: boolean;
  isCityTraffic: boolean;
  tripDistanceKm: number;
  tripDurationS: number;
  isVehicleStopped: boolean;
  sensitivityOffsetC: number;
  liveEcoScore: number;
}

export function emptyCareContext(now = Date.now()): CareContext {
  return {
    now,
    vehicle: archetypeProfile("gasolineNA", "gasoline"),
    oilIsEstimated: true,
    engineReady: false,
    isColdPhase: true,
    isCityTraffic: false,
    tripDistanceKm: 0,
    tripDurationS: 0,
    isVehicleStopped: true,
    sensitivityOffsetC: 0,
    liveEcoScore: 100,
  };
}

export function bucketAmbient(c: number | undefined): string {
  if (c == null) return "ambient:unknown";
  if (c < 0) return "ambient:<0";
  if (c < 10) return "ambient:0-10";
  if (c < 20) return "ambient:10-20";
  if (c < 30) return "ambient:20-30";
  return "ambient:>30";
}

export function bucketLoad(pct: number | undefined): string {
  if (pct == null) return "load:unknown";
  if (pct < 40) return "load:<40";
  if (pct < 70) return "load:40-70";
  return "load:>70";
}

export function bucketSpeed(kmh: number | undefined): string {
  if (kmh == null) return "speed:unknown";
  if (kmh < 0.5) return "speed:0";
  if (kmh < 50) return "speed:1-50";
  if (kmh < 90) return "speed:50-90";
  return "speed:>90";
}

export function compositeBucket(ambient: number | undefined, load: number | undefined, speed: number | undefined): string {
  return [bucketAmbient(ambient), bucketLoad(load), bucketSpeed(speed)].join("|");
}
